src/test/java/com/puppycrawl/tools/checkstyle/checks/header/MultiFileRegexpHeaderCheckTest.java

```diff
@@ -0,0 +1,769 @@
+        checkConfig.addProperty("headerFiles", headerFile.getPath());
+
+        final String[] expected = {
+            "1: " + getCheckMessage(MSG_HEADER_MISMATCH, "// Header content"),
+        };
+        verify(checkConfig, testFile.getPath(), expected);
+    }
+
+    @Test
+    public void testNonEmptyPatternsWithMismatchedCardinality() throws Exception {
+        // Create a test file with content that will match some but not all patterns
+        final File testFile = new File(temporaryFolder, "test.java");
+        Files.write(testFile.toPath(), List.of(
+            "// First line matches",
+            "// Second line matches",
+            "// Third line doesn't match pattern"), StandardCharsets.UTF_8);
+
+        // Create a header file with patterns that should match the first two lines
+        // but not the third line
+        final File headerFile = new File(temporaryFolder, "header.header");
+        Files.write(headerFile.toPath(), List.of(
+            "// First line matches",
+            "// Second line matches",
+            "// Third line with different pattern"), StandardCharsets.UTF_8);
+
+        final DefaultConfiguration checkConfig =
+                createModuleConfig(MultiFileRegexpHeaderCheck.class);
+        checkConfig.addProperty("headerFiles", headerFile.getPath());
+
+        final String[] expected = {
+            "3: " + getCheckMessage(MSG_HEADER_MISMATCH, "// Third line with different pattern"),
+        };
+        verify(checkConfig, testFile.getPath(), expected);
+    }
+
+    @Test
+    public void testFileSizeLessThanHeaderPatternSize() throws Exception {
+        // Create a header file with more patterns than the file has lines
+        final File headerFile = new File(temporaryFolder, "longerHeader.header");
+        Files.write(headerFile.toPath(), List.of(
+            "// First line",
+            "// Second line",
+            "// Third line",
+            "// Fourth line",
+            "// Fifth line"), StandardCharsets.UTF_8);
+
+        // Create a test file with fewer lines than the header
+        final File testFile = new File(temporaryFolder, "shorterFile.java");
+        Files.write(testFile.toPath(), List.of(
+            "// First line",
+            "// Second line",
+            "// Third line"), StandardCharsets.UTF_8);
+
+        final DefaultConfiguration checkConfig =
+                createModuleConfig(MultiFileRegexpHeaderCheck.class);
+        checkConfig.addProperty("headerFiles", headerFile.getPath());
+
+        // File should fail because it has fewer lines than the header patterns
+        final String[] expected = {
+            "1: " + getCheckMessage(MSG_HEADER_MISSING),
+        };
+        verify(checkConfig, testFile.getPath(), expected);
+
+        // Now create another header with blank patterns at the end
+        final File headerWithBlanksFile = new File(temporaryFolder, "headerWithBlanks.header");
```

Comment:
please split in two methods, we keep single verify in test method, is easier to deal with.